contracts/src/PDPService.sol

```diff
@@ -253,31 +284,36 @@ contract PDPService {
         return rootId;
     }
 
+    function scheduleRemovals(uint256 setId, uint256[] calldata rootIds) public {
+        require(rootIds.length + scheduledRemovals[setId].length <= MAX_ENQUEUED_REMOVALS, "Too many removals wait for next proving period to schedule");
+        require(proofSetOwner[setId] == msg.sender, "Only the owner can schedule removal of roots");
+        require(proofSetLive(setId), "Proof set not live");
+        uint256 totalDelta = 0;
+        for (uint256 i = 0; i < rootIds.length; i++){
+            totalDelta += rootLeafCounts[setId][rootIds[i]];
+            scheduledRemovals[setId].push(rootIds[i]);
+        }
+        
+        bytes memory extraData = abi.encode(totalDelta, rootIds);
+        _notifyListener(setId, proofSetListener[setId], PDPListener.OperationType.REMOVE_SCHEDULED, extraData);
+    }
+
     // removeRoots removes a batch of roots from a proof set.  Must be called by the proof set owner.
     // returns the total removed leaf count
-    function removeRoots(uint256 setId, uint256[] calldata rootIds) public returns (uint256){
-        require(proofSetOwner[setId] == msg.sender, "Only the owner can remove roots");
+    function removeRoots(uint256 setId, uint256[] memory rootIds) internal returns (uint256){
         require(proofSetLive(setId), "Proof set not live");
         uint256 totalDelta = 0;
         for (uint256 i = 0; i < rootIds.length; i++){
             totalDelta += removeOneRoot(setId, rootIds[i]);
         }
         proofSetLeafCount[setId] -= totalDelta;
-        // Clear next challenge epoch if the set is now empty.
-        // It will be re-set when new data is added.
-        if (proofSetLeafCount[setId] == 0) {
-            nextChallengeEpoch[setId] = 0;
-        }
-
-        bytes memory extraData = abi.encode(totalDelta, rootIds);
-        _notifyListener(setId, proofSetListener[setId], PDPListener.OperationType.REMOVE, extraData);
-        emit RootsRemoved(totalDelta);
 
         return totalDelta;
     }
 
     // removeOneRoot removes a root from a proof set. 
     function removeOneRoot(uint256 setId, uint256 rootId) internal returns (uint256) {
+        emit Debug("removing root", rootId);
```

Comment:
Remove the debug before merging?

Reply:
Thanks
Off topic but I just discovered this way of debugging solidity contracts and it works really nicely